pkg/fileutil: lock file on Windows

gyuho opened the pull request:
Fix the following:

1. windows: fix WAL lock https://github.com/coreos/etcd/issues/1972
2. etcdserver: failed to purge snap file https://github.com/coreos/etcd/issues/2515

Tested in Windows machine and tests passed.

/cc @heyitsanthony @xiang90 

Thanks.

heyitsanthony commented on pkg/fileutil/readdir_windows.go:
```diff
@@ -0,0 +1,46 @@
+	"sort"
+)
+
+func ReadDir(dir string) ([]string, error) {
```
Is windows generating extra files? NFS has an issue with `.nfs` lock files...

Both windows and linux should probably do the same thing here-- get the directory file names from `os.Readdirnames(-1)` and filter out the dot files.

gyuho replied:
Yeah it was reading the locked file even after it deletes that file.  I will double check. Thanks

heyitsanthony commented on pkg/fileutil/lock_windows.go:
```diff
@@ -16,17 +16,112 @@
+	if fd == syscall.InvalidHandle {
+		return nil
+	}
+	for {
```
why does this need a retry loop?

heyitsanthony commented on pkg/fileutil/fileutil_test.go:
```diff
@@ -41,10 +42,10 @@ func TestIsDirWriteable(t *testing.T) {
 		t.Skipf("failed to get current user: %v", err)
 	}
-	if me.Name == "root" || me.Name == "Administrator" {
+	if me.Name == "root" || me.Name == "Administrator" || runtime.GOOS == "windows" {
```
`me.Name == "root" || runtime.GOOS == "windows" `?

gyuho commented:
@heyitsanthony PTAL.

1. removed `time.Sleep` (not needed). I tested without it and worked.
2. added comment about why we need separate `ReadDir` function for Windows, which returns outdated files in some systems (it's `nil`)
3. removed `Administrator` checking in `fileutil_test.go`. Windows doesn't support Chmod, so we just skip it.

Thanks!

heyitsanthony commented on pkg/fileutil/readdir_windows.go:
```diff
@@ -0,0 +1,48 @@
+		}
+		return nil
+	}
+	if err := filepath.Walk(dir, visit); err != nil {
```
Would `Readdir` work instead? `Walk` seems somewhat extreme.

gyuho replied:
I tried `Readdir`, basically same as `Readdirnames` (which uses `Readdir`) and it returns `nil` files in Windows as well.

heyitsanthony replied:
OK, I looked at the go sources to figure out what's happening here. It seems the main difference between `readdirnames`/`readdir` and `walk` is `walk` is `lstat`ing the files after it gets the names so there could be some race happening and this is masking it.

Is this so the PurgeFile tests pass or is something else failing too?

gyuho replied:
Only this fails https://github.com/coreos/etcd/blob/master/pkg/fileutil/purge_test.go#L146. `5.test` is `nil` (I checked manually as well), but it is returned in `Readdir`.

heyitsanthony commented on pkg/fileutil/lock_windows.go:
```diff
@@ -16,17 +16,106 @@
 		return nil, err
 	}
+	if err := lockFile(syscall.Handle(f.Fd()), 0); err != nil {
+		return nil, err
```
`f.Close()`?

heyitsanthony commented on pkg/fileutil/lock_windows.go:
```diff
@@ -16,17 +16,106 @@
+		return nil, err
+	}
+	if err := lockFile(syscall.Handle(f.Fd()), LOCKFILE_FAIL_IMMEDIATELY); err != nil {
+		return nil, err
```
`f.Close()`?

gyuho commented on pkg/fileutil/lock_windows.go:
```diff
@@ -16,17 +16,108 @@
+		return nil, err
+	}
+	if err := lockFile(syscall.Handle(f.Fd()), LOCKFILE_FAIL_IMMEDIATELY); err != nil {
+		f.Close()
```
@heyitsanthony  Just added `Close`s. Thanks.

heyitsanthony replied:
This didn't happen to fix the readdir thing, did it? Windows might be picky about fd's hanging around.

heyitsanthony commented on pkg/fileutil/lock_windows.go:
```diff
@@ -16,17 +16,106 @@
+
+func lockFile(fd syscall.Handle, flags uint32) error {
+	var flag uint32 = LOCKFILE_EXCLUSIVE_LOCK
+	if flags != 0 {
```
unnecessary `if`?

gyuho replied:
you are right. Just removed. Thanks!

gyuho commented:
@heyitsanthony wow you are totally right. After adding `f.Close`, tests pass. Just fixed back to use `ReadDir`. Thanks!

heyitsanthony commented on pkg/fileutil/lock_windows.go:
```diff
@@ -16,17 +16,106 @@
+		return nil, fmt.Errorf("cannot open empty filename")
+	}
+	var access uint32
+	switch flag & (syscall.O_RDONLY | syscall.O_WRONLY | syscall.O_RDWR) {
```
`panic` on  `default:` instead of masking out flags?

heyitsanthony replied:
This should be `switch flag {`...?

heyitsanthony commented:
OK lgtm aside from the open flag stuff. Thanks!

gyuho commented on pkg/fileutil/lock_windows.go:
```diff
@@ -16,17 +16,108 @@
+	case syscall.O_RDWR:
+		access = syscall.GENERIC_READ | syscall.GENERIC_WRITE
+	default:
+		panic(fmt.Errorf("flag %v is not supported", flag))
```
just added. Will merge after green lights. Thanks!!

gyuho replied:
@heyitsanthony just fixed to `switch flag`. Thanks for double-checking!

heyitsanthony replied:
@gyuho you may want to confirm `etcd` runs as expected on windows. There's an `O_CREATE` flag that may cause this to panic (`flag & O_CREATE`, I guess?)

gyuho replied:
Yeah actually just tested and found the panic. So I added another check with `syscall.O_WRONLY | syscall.O_CREAT`. Now works.

Thanks!
